### PR TITLE
fix(@aws-amplify/datastore): extra fields in selective sync selection set on API's from older CLI

### PR DESCRIPTION
**Pull request by svidgen**
#### Description of changes

Fixes an issue where the selection set for selective sync expressions was incorrect for apps built with some older versions of the CLI. Related-model key fields were included in the selection set, but were not recognized by cloud storage. This issue was introduced in with the nested predicates feature, which added the implicitly created FK fields to the table metadata `fields` list so they could be seen by the predicate builder.

This PR removes the logic that augments the `fields` list and add a new `allFields` list. All pre-existing logic and components (prior to lazy loading/nested predicates) will revert to use only the explicitly defined `fields` list. The predicates builder will then use `allFields`.

~To support this, storage predicate construction is refactored slightly to avoid using the storage predicate (v1 predicate) builder. The old builder complains when a field from `allFields` shows up, even though the field exist. So, we are instead constructing an AST and transforming it.~

~The resulting storage predicate can, at times, contain joins against `null` or `undefined` values. This happens when a relationship is optional. So, the IndexedDBAdapter was also updated to avoid attempting a `index.get()` on these values, which otherwise causes IndexedDB to throw a fit.~

(Refactors around storage predicate generation and related IndexedDB fixes were handled upstream.)

#### Issue #, if available

#10785

#### Description of how you validated changes

1. Added testing.
4. Manual sample apps validation: Red/Green'd `many-to-many` app with added selective syncs. [That PR](https://github.com/aws-amplify/amplify-js-samples-staging/pull/530) can be merged after this one.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/datastore/src/predicates/next.ts [hunk 1]
```diff
@@ -169,6 +187,37 @@ export class FieldCondition {
...
+	 * @param model The model meta to use when construction a new `GroupCondition`
+	 * for cases where the negation requires multiple `FieldCondition`'s.
+	 */
+	negated(model: ModelMeta<any>) {
+		if (this.operator === 'between') {
+			return new GroupCondition(model, undefined, undefined, 'or', [
+				new FieldCondition(this.field, 'lt', [this.operands[0]]),
+				new FieldCondition(this.field, 'gt', [this.operands[1]]),
+			]);
+		} else {
```
Comment (dpilch) [inline packages/datastore/src/predicates/next.ts:212]: I think I missed this on a previous review. What about `beginsWith`?

Comment (svidgen) [reply packages/datastore/src/predicates/next.ts:212]: Looks like an existing bug. I'm adding a bunch of negation tests and will fix this in another PR ASAP.

Comment (svidgen) [reply packages/datastore/src/predicates/next.ts:212]: https://github.com/aws-amplify/amplify-js/issues/10957

File: packages/datastore/src/storage/adapter/IndexedDBAdapter.ts [hunk 2]
```diff
@@ -443,6 +443,17 @@ class IndexedDBAdapter implements Adapter {
 		const hasPagination = pagination && pagination.limit;
 
 		const records: T[] = (await (async () => {
+			//
```
Comment (david-mcafee) [inline packages/datastore/src/storage/adapter/IndexedDBAdapter.ts:446]: What about the other storage adapters? Is only IDB affected?

File: packages/datastore/src/predicates/next.ts (hunk 1 above)
Comment (svidgen) [reply packages/datastore/src/predicates/next.ts:212]: Fix: https://github.com/aws-amplify/amplify-js/pull/10962

Comment (svidgen) [reply packages/datastore/src/predicates/next.ts]: Fixed upstream.

File: packages/datastore/src/storage/adapter/IndexedDBAdapter.ts (hunk 2 above)
Comment (svidgen) [reply packages/datastore/src/storage/adapter/IndexedDBAdapter.ts]: Discussed offline and incorporated upstream. But, for posterity, these changes to IndexedDBAdapter handle idiosyncrasies specific to IndexedDB, which throws a fit if you try to `index.get([null])`. The changes here would prevent that query pattern, and force execution through an `index.queryAll(<predicate>)` line of execution instead.

**Review by dpilch — APPROVED** [review]

**Review by iartemiev — APPROVED** [review]

**Pull request merged**